### PR TITLE
Nested Combobox Grouping

### PR DESCRIPTION
**Pull request by awseward**
This PR enables the `Combobox` to group items in an n-tiered fashion.

**Note**: Please ignore all changes made where are not under the `src/` directory.
### Before

screen shot 2016-04-19 at 15 32 14
### After

screen shot 2016-04-19 at 15 28 02

### CODE REVIEW

File: src/GroupHeader.jsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import React, { PropTypes } from 'react';
+import CustomPropTypes from './util/propTypes';
+
+function GroupHeader(props) {
+  const {
```
Comment (yoyis3000) [inline src/GroupHeader.jsx:5]: NB: Why not get className and id while you're at it?

File: src/Combobox.jsx [hunk 2]
```diff
@@ -148,7 +155,7 @@ var ComboBox = React.createClass({
       , afterListComponent, searchTerm, onChange
       , listComponent: List } = this.props;
 
-    List = List || (groupBy && GroupableList) || PlainList
+    List = List || _resolveListGroupable(groupBy) || PlainList
```
Comment (kespinola) [inline src/Combobox.jsx]: `DropdownList` and `Combobox` take a `listComponent` prop. What do you think of exposing `MultiGroupList` from the package then passing `MultiGroupList` to `listCompoent` when you instaniate a new `Combobox` like so

```
import { MultiGroupList, Combobox } from 'react-widgets';

render() {
  <Combobox listComponent={MultGroupList} />
}
```

Naturally extending the functionality using the given api instead of adding checks for our special case.

**Comment by yoyis3000** [comment]
+1

File: src/Combobox.jsx (hunk 2 above)
Comment (awseward) [inline src/Combobox.jsx]: Oh I like that a lot. Awesome, thanks!

File: src/GroupHeader.jsx (hunk 1 above)
Comment (awseward) [inline src/GroupHeader.jsx:5]: @yoyis3000 I usually try to hold off on destructuring things until I have to reference them more than once in the enclosing scope.

File: src/ListMultiGroupable.jsx [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+  },
+
+  getItemDOMNode(item) {
+    const list = compat.findDOMNode(this);
+    const index = _getOrderedIndex(item, this.state.groups);
+
+    // Conveniently, someArray[-1] gives undefined, which is just what we want
+    return list[index];
+  }
+});
```
Comment (kespinola) [inline src/ListMultiGroupable.jsx:409]: Mind giving me a run down of this when you get in? Thanks.

Comment (awseward) [inline src/ListMultiGroupable.jsx:409]: I'm in now, so anytime before 9:30 is good for me.

File: lib/ListGroupable.js [hunk 4]
```diff
@@ -236,8 +236,7 @@ exports['default'] = _react2['default'].createClass({
 
   _data: function _data() {
     var groups = this.state.groups;
-
-    return this.state.sortedKeys.reduce(function (flat, grp) {
+    return result = this.state.sortedKeys.reduce(function (flat, grp) {
```
Comment (farism) [inline lib/ListGroupable.js]: n/b: Unnecessary whitespace diff

File: lib/index.js [hunk 5]
```diff
@@ -11,12 +11,13 @@ if (process.env.NODE_ENV !== 'production') {
...
-  DropdownList: require('./DropdownList'),
-  Combobox: require('./Combobox'),
   Calendar: require('./Calendar'),
+  Combobox: require('./Combobox'),
   DateTimePicker: require('./DateTimePicker'),
-  NumberPicker: require('./NumberPicker'),
+  DropdownList: require('./DropdownList'),
+  ListMultiGroupable: require('./ListMultiGroupable'),
   Multiselect: require('./Multiselect'),
+  NumberPicker: require('./NumberPicker'),
```
Comment (farism) [inline lib/index.js]: n/b: Same as above w/ whitespace, if we are going to hack on 3rd party modules we should only do the minimum required changes.

File: lib/less/core.less [hunk 6]
```diff
@@ -211,15 +211,34 @@ ul.rw-list {
...
+        }
+
+        &-4 {
+          padding-left: 50px;
+        }
+
+        &-5 {
+          padding-left: 60px;
+        }
+      }
```
Comment (farism) [inline lib/less/core.less]: I think you're looking for the `@for` control directive here

```
@for $i from 0 through 5
  &-#{$i}
    width: ($i * 10) + 10px
```

or something like that.

File: lib/less/core.less [hunk 7]
```diff
@@ -211,15 +211,34 @@ ul.rw-list {
...
-
-    > li.rw-list-optgroup {
-      padding-left: 10px;
-    }
-
-    > li.rw-list-option {
-      padding-left: 20px;
+    > li {
+      &.rw-list-depth {
+        // TODO: There's probably some kind of expression syntax for this...
```
Comment (farism) [inline lib/less/core.less]: I think you're looking for the `@for` control directive here

```
@for $i from 0 through 5
  &-#{$i}
    width: ($i * 10) + 10px
```

or something like that.

File: src/less/core.less [hunk 8]
```diff
@@ -211,15 +211,34 @@ ul.rw-list {
...
-
-    > li.rw-list-optgroup {
-      padding-left: 10px;
-    }
-
-    > li.rw-list-option {
-      padding-left: 20px;
+    > li {
+      &.rw-list-depth {
+        // TODO: There's probably some kind of expression syntax for this...
```
Comment (farism) [inline src/less/core.less:217]: loops in LESS are kind of gross

File: src/index.js [hunk 9]
```diff
@@ -13,13 +13,14 @@ if (process.env.NODE_ENV !== 'production' ) {
...
-  Multiselect:      require('./Multiselect'),
-  SelectList:       require('./SelectList'),
+  Calendar:            require('./Calendar'),
+  Combobox:            require('./Combobox'),
+  DateTimePicker:      require('./DateTimePicker'),
+  DropdownList:        require('./DropdownList'),
+  ListMultiGroupable:  require('./ListMultiGroupable'),
+  Multiselect:         require('./Multiselect'),
+  NumberPicker:        require('./NumberPicker'),
+  SelectList:          require('./SelectList'),
```
Comment (farism) [inline src/index.js]: n/b: Same as above w/ whitespace, if we are going to hack on 3rd party modules we should only do the minimum required changes.

File: src/ListGroupable.jsx [hunk 10]
```diff
@@ -210,9 +210,8 @@ export default React.createClass({
 
   _data(){
     var groups = this.state.groups;
-
-    return this.state.sortedKeys
-      .reduce( (flat, grp) => flat.concat(groups[grp]), [])
+    return result = this.state.sortedKeys
```
Comment (farism) [inline src/ListGroupable.jsx]: n/b: Unnecessary whitespace diff

File: package.json [hunk 11]
```diff
@@ -45,6 +45,7 @@
     "build:docs": "npm run clean:docs && npm run compile:docs && npm run assets:docs",
     "build:dist": "npm run clean:dist && npm run compile:dist & npm run compile:locale && npm run less && npm run assets:dist",
     "build": "npm run build:lib && npm run build:dist & npm run build:docs",
+    "b": "npm run build:lib && npm run build:dist",
```
Comment (farism) [inline package.json]: Do we need this shortcut run script in our fork?

Comment (awseward) [inline package.json]: @farism Good catch, forgot to remove this.

File: src/ListMultiGroupable.jsx [hunk 12]
```diff
@@ -0,0 +1,477 @@
...
+
+function _pushNewOrderedKey(obj, key) {
+  const shouldPushKey = obj
+    && obj._orderedKeys
+    && obj._orderedKeys.indexOf(key) === -1;
+
+  shouldPushKey && obj._orderedKeys.push(key);
+}
+
+function _setIn(obj, path, val) {
```
Comment (remyyounes) [inline src/ListMultiGroupable.jsx:42]: Switching your arguments to have obj always last would allow for FP.

**Comment by kespinola** [comment]
+1

File: lib/ListMultiGroupable.js [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+      outputArray.push(renderSingleItem(item, state, idx));
+    });
+  };
+
+  _utilObjectTraversal.depthFirst(groupedObj, getChildren, onInternal, onLeaf);
+
+  return outputArray;
+}
+
+function _renderHeadersAndItems(groupedData, processHeader, processItems) {
```
Comment (remyyounes) [inline lib/ListMultiGroupable.js]: where are these being called ?

Comment (awseward) [inline lib/ListMultiGroupable.js]: Some of these are old functions I forgot to pull out.

Thanks for catching them

File: src/ListMultiGroupable.jsx (hunk 12 above)
Comment (awseward) [inline src/ListMultiGroupable.jsx:42]: Ooh I like. May do this in a followup

**Comment by remyyounes** [comment]
+1

**Comment by thomasm-procore** [comment]
Passed testing.
